base/reduce.jl

```diff
@@ -103,30 +103,21 @@ julia> foldl(=>, 1:4)
 foldl(op, itr) = mapfoldl(identity, op, itr)
 
 ## foldr & mapfoldr
-
-function mapfoldr_impl(f, op, v0, itr, i::Integer)
-    # Unroll the while loop once; if v0 is known, the call to op may
-    # be evaluated at compile time
-    if isempty(itr) || i == 0
-        return v0
-    else
-        x = itr[i]
-        v  = op(f(x), v0)
-        while i > 1
-            x = itr[i -= 1]
-            v = op(f(x), v)
-        end
-        return v
-    end
+# switches the order of arguments of a binary operator
+struct Switch{F}
+    op::F
 end
+(s::Switch)(x,y) = s.op(y,x)
+reduce_empty(s::Switch, T) = reduce_empty(s.op, T)
+reduce_first(s::Switch, x) = reduce_first(s.op, x)
```

Comment:
Though this is elegant, it doesn't cover the specialized cases of `mapreduce_empty_iter` that we already have, not to mention future specializations of `mapreduce_first`. 

You can try to override `mapreduce_empty_iter` and `mapreduce_first`, but that seems to be a minefield of potential method ambiguities.